patch/0125-devlink-add-support-to-create-line-card-and-expose-t.patch

```diff
@@ -0,0 +1,532 @@
++				    enum devlink_command cmd)
++{
++	struct devlink *devlink = linecard->devlink;
++	struct sk_buff *msg;
++	int err;
++
++	WARN_ON(cmd != DEVLINK_CMD_LINECARD_NEW &&
++		cmd != DEVLINK_CMD_LINECARD_DEL);
++
++	if (!xa_get_mark(&devlinks, devlink->index, DEVLINK_REGISTERED))
++		return;
++
++	msg = nlmsg_new(NLMSG_DEFAULT_SIZE, GFP_KERNEL);
++	if (!msg)
++		return;
++
++	err = devlink_nl_linecard_fill(msg, devlink, linecard, cmd, 0, 0, 0,
++				       NULL);
++	if (err) {
++		nlmsg_free(msg);
++		return;
++	}
++
++	genlmsg_multicast_netns(&devlink_nl_family, devlink_net(devlink),
++				msg, 0, DEVLINK_MCGRP_CONFIG, GFP_KERNEL);
++}
++
++static int devlink_nl_cmd_linecard_get_doit(struct sk_buff *skb,
++					    struct genl_info *info)
++{
++	struct devlink_linecard *linecard = info->user_ptr[1];
++	struct devlink *devlink = linecard->devlink;
++	struct sk_buff *msg;
++	int err;
++
++	msg = nlmsg_new(NLMSG_DEFAULT_SIZE, GFP_KERNEL);
++	if (!msg)
++		return -ENOMEM;
++
++	err = devlink_nl_linecard_fill(msg, devlink, linecard,
++				       DEVLINK_CMD_LINECARD_NEW,
++				       info->snd_portid, info->snd_seq, 0,
++				       info->extack);
++	if (err) {
++		nlmsg_free(msg);
++		return err;
++	}
++
++	return genlmsg_reply(msg, info);
++}
++
++static int devlink_nl_cmd_linecard_get_dumpit(struct sk_buff *msg,
++					      struct netlink_callback *cb)
++{
++	struct devlink_linecard *linecard;
++	struct devlink *devlink;
++	int start = cb->args[0];
++	unsigned long index;
++	int idx = 0;
++	int err;
++
++	mutex_lock(&devlink_mutex);
++	xa_for_each_marked(&devlinks, index, devlink, DEVLINK_REGISTERED) {
++		if (!devlink_try_get(devlink))
++			continue;
++
++		if (!net_eq(devlink_net(devlink), sock_net(msg->sk)))
++			goto retry;
++
++		mutex_lock(&devlink->linecards_lock);
++		list_for_each_entry(linecard, &devlink->linecard_list, list) {
++			if (idx < start) {
++				idx++;
++				continue;
++			}
++			err = devlink_nl_linecard_fill(msg, devlink, linecard,
++						       DEVLINK_CMD_LINECARD_NEW,
++						       NETLINK_CB(cb->skb).portid,
++						       cb->nlh->nlmsg_seq,
++						       NLM_F_MULTI,
++						       cb->extack);
++			if (err) {
++				mutex_unlock(&devlink->linecards_lock);
++				devlink_put(devlink);
++				goto out;
++			}
++			idx++;
++		}
++		mutex_unlock(&devlink->linecards_lock);
++retry:
++		devlink_put(devlink);
++	}
++out:
++	mutex_unlock(&devlink_mutex);
++
++	cb->args[0] = idx;
++	return msg->len;
++}
++
+ static int devlink_nl_sb_fill(struct sk_buff *msg, struct devlink *devlink,
+ 			      struct devlink_sb *devlink_sb,
+ 			      enum devlink_command cmd, u32 portid,
+@@ -7594,6 +7818,19 @@ static const struct nla_policy devlink_nl_policy[DEVLINK_ATTR_MAX + 1] = {
+ 	[DEVLINK_ATTR_RELOAD_ACTION] = NLA_POLICY_RANGE(NLA_U8, DEVLINK_RELOAD_ACTION_DRIVER_REINIT,
+ 							DEVLINK_RELOAD_ACTION_MAX),
+ 	[DEVLINK_ATTR_RELOAD_LIMITS] = NLA_POLICY_BITFIELD32(DEVLINK_RELOAD_LIMITS_VALID_MASK),
++//<<<<<<< HEAD
```

Comment:
Can these conflict markers be removed? They're comments and so won't have any impact, but it would be nice to remove them.

Reply:
@saiarcot895 This is removed in patch 0126. Hence this will not be in the final code. However I get your concern about the markers and we will address this in the next update which is planned soon.

Reply:
Resolved in latest commit